src/metabase/db/liquibase.clj

```diff
@@ -181,20 +189,44 @@
   see https://github.com/metabase/metabase/issues/3715"
   [db-type :- s/Keyword
    conn    :- java.sql.Connection]
-  (let [liquibase-table-name (if (#{:h2 :mysql} db-type)
-                               "DATABASECHANGELOG"
-                               "databasechangelog")
+  (let [liquibase-table-name (changelog-table-name db-type)
         fresh-install?       (jdbc/with-db-metadata [meta {:connection conn}] ; don't migrate on fresh install
                                (empty? (jdbc/metadata-query
                                         (.getTables meta nil nil liquibase-table-name (u/varargs String ["TABLE"])))))
         statement            (format "UPDATE %s SET FILENAME = ?" liquibase-table-name)]
     (when-not fresh-install?
       (jdbc/execute! {:connection conn} [statement "migrations/000_migrations.yaml"]))))
 
-(defn rollback-one
-  "Roll back the last migration."
-  [^Liquibase liquibase]
-  (.rollback liquibase 1 ""))
+(defn- extract-numbers
+  "Returns contiguous integers parsed from string s"
+  [s]
+  (map #(Integer/parseInt %) (re-seq #"\d+" s)))
+
+(defn- current-major-version
+  "Returns the major version of the running Metabase JAR"
+  []
+  (second (extract-numbers (:tag config/mb-version-info))))
+
+(defn rollback-major-version
+  "Roll back migrations later than given Metabase major version"
+  ;; default rollback to previous version
+  ([db-type conn liquibase]
+   ;; get current major version of Metabase we are running
+   (rollback-major-version db-type conn liquibase (dec (current-major-version))))
+
+  ;; with explicit target version
+  ([db-type conn ^Liquibase liquibase target-version]
+   (when (or (not (integer? target-version)) (< target-version 44))
+     (throw (IllegalArgumentException.
+             (trs "target version must be a number between 44 and the previous major version ({0}), inclusive" (current-major-version)))))
+   ;; count and rollback only the applied change set ids which come after the target version (only the "v..." IDs need to be considered)
+   (let [changeset-query (format "SELECT id FROM %s WHERE id LIKE 'v%%' ORDER BY ORDEREXECUTED ASC"
```

Comment:
I'm surprised this works on MySQL without quoting identifiers since for "reasons" we have the table named as `DATABASCHANGELOG` (all caps) and MySQL lowercases unquoted identifiers by default. Are the tests passing for MySQL? Maybe our tests are busted if they are.

Reply:
I've tested with MariaDB, and the rollback test succeeds on all drivers in CI.